notebooks/circuits_with_bloqade.py

```diff
@@ -0,0 +1,763 @@
+    squin.op.reset() # Reset a qubit to the |0> state
+    squin.op.phase(theta=0.1) # Add a global phase to the qubits
+    # Transforming gates into new operators:
+    squin.op.control(X,n_controls=2) # An X gate conditional on 2 qubits, aka a Toffoli gate
+    squin.op.rot(axis=X, angle=0.5) # A rotation around the X axis by 0.5 radians
+    squin.op.adjoint(t) # The inverse of a gate
+    
+    # Operators that are not unitaries to represent Paulis
+    squin.op.scale(op=X, factor=0.1) # Scale an operator by some factor, eg 0.1*X
+    squin.op.kron(lhs=X,rhs=Y) # The Kroneker product of two operators to act on multiple qubits, eg X ⊗ Y
+    squin.op.mult # Is this necessary?
+    squin.op.identity # The identity operator, which does nothing
+    squin.op.pauli_string(string="XXYZ") # A simpler alias for squin.op.kron(squin.op.X(), ...)
+
+
+# %% [markdown]
+# For example, lets make some operators and gates that go beyond the basic operators. Note that more complicated functionalities, such as T state teleportation, need function signatures that do not match that of the builtin `squin.gate` functions, and might need to be written using `op`. For example, consider a kernel that applies a controlled T gate to a register of qubits
+
+# %%
+try:
+    del controlled_t
+except:
+    pass
+try:
+    del heisenberg_rotation
+except:
+    pass
+@squin.kernel
+def controlled_t(qubits:Register):
+    """
+    A controlled T gate, aka a sqrt{CZ} gate.
+    ---o---
+       |
+    ---T---
+    """
+    t = squin.op.t()
+    ctrl = squin.op.control(t, n_controls=1)
+    squin.qubit.apply(operator=ctrl, qubits=qubits)
+
+controlled_t.print()
+
+# %% [markdown]
+# Finally, you can also define noise on qubits using the `squin.noise` namespace. Like `op`, `noise` operators are defined separately from their application to qubits
+
+# %%
+try:
+    del _noise
+except:
+    pass
+@squin.kernel
+def _noise():
+
+    X = squin.op.x()
+    squin.noise.pauli_error(basis=X, p=0.1) # An X flip with probability 0.1
+    loss = squin.noise.qubit_loss(p=0.1) # A qubit loss with probability 0.1
+    squin.noise.depolarize(p=0.1) # Depolarizing channel https://quantumai.google/reference/python/cirq/DepolarizingChannel
+    
+    squin.noise.single_qubit_pauli_channel # ???
+    squin.noise.two_qubit_pauli_channel # ???
+    
+    # Noise is applied to qubits in the same way as operators.
+    qubit = squin.qubit.new(1)[0]
+    squin.qubit.apply(operator=loss, qubits=[qubit])
+
+# %% [markdown]
+# ## Using Bloqade kernels
+# A key feature of kernels is the ability to do complex control flow similar to how one might program python. For example, one can use a for loop to apply the same gate to multiple qubits to prepare a GHZ state.
+# 
+# A common pattern is to use factory functions that return bloqade kernels. This way, you can fix parameters (such as the number of qubits) pythonically without needing to introduce the variable directly into the kernel itself.
```

Comment:
To be honest, I'm not sure I agree that this is a "common" pattern 😄 For starters, given the state of squin in general, I'd say it's a bit soon to call something common. Also, why don't you just do

```python
def GHZ_state(nqubits: int):
        qubits = squin.qubit.new(nqubits)
        squin.gate.h(qubits[0])
        for i in range(nqubits):
            squin.gate.cx(qubits[i], qubits[i+1])

        return qubits
```

That also works just fine and is a bit clearer IMO. You then just need to pass in the parameter as argument when running an interpreter such as the simulator.